### DISCUSSION TITLE
How to connect Button Matrix?

### DISCUSSION THREAD

**Discussion by kentforth** [tttapa/Control-Surface#441] [General]
Hey Pieter!
I have 4 buttons. I want to use button matrix . One example for button notes and one example for buttons midi cc
I tried to connect buttons like this but it doesn't work. 
**How to connect buttons in matrix properly?**

**Can you show me sketch examples with notes and midi cc?**

![Button Matrix Connection](https://user-images.githubusercontent.com/17134421/112727842-f4aefc00-8f56-11eb-8a42-ef44f4a524ec.jpg)

```
#include <Control_Surface.h>

USBMIDI_Interface midi;

// The note numbers corresponding to the buttons in the matrix
const AddressMatrix<2, 2> addresses = {{
  { 1, 2 },
  { 3, 4}, 
}};

NoteButtonMatrix<2, 2> buttonmatrix = {
  {2, 3}, 
  {4, 5}, 
  addresses, 
  CHANNEL_1,
};

void setup() {
  Control_Surface.begin();
}

void loop() {
  Control_Surface.loop();
}
```

**Comment by tttapa** [comment]
Your matrix should not be connected to Vcc. See the schematic here: https://tttapa.github.io/Control-Surface-doc/new-input/Doxygen/d4/d53/NoteButtonMatrix_8ino-example.html

For 4 buttons, there is nothing to be gained from a matrix layout. Just connect each button between an input pin and ground.

**Reply by kentforth** [reply]
What is the purpose of button matrix if I connect each button separately? I mean the button matrix should require less arduino pins than each button separately right?

**Reply by tttapa** [reply]
The advantage of button matrices is that they scale better in function of the number of buttons. If you connect each of N buttons separately, you need exactly N pins. If you connect them in a matrix, you need around 2×√N pins. 

In your case, N = 4, so to connect them separately, you need N = 4 pins, and to connect them in a matrix, you need 2×√4 = 4 pins, so it makes no sense to use the matrix approach here, it's only useful for N > 5.

**Reply by kentforth** [reply]
I don't understand how should I connect buttons in  button matrix if  have more than 4 buttons. Let's say I will have 10 buttons. If I connect each button to each pin on arduino the button matrix doesn't make sense for me. Can you explan this thing?

**Reply by tttapa** [reply]
I don't understand what you mean. If you have 5 or fewer buttons, don't use a matrix, each button connects to one Arduino pin.

If you have more buttons, you can use the matrix schematic I posted in my first reply.

**Reply by kentforth** [reply]
The schematics shows that button in column should be connected to +5v through resistor, but you told me to connect a button between arduino pin and GND,
I'm frustrated. How should I connect buttons properly?

**Reply by tttapa** [reply]
No, there are two mutually exclusive options: 

1. Connect each button to its own pin. In that case, connect one lead of each button to ground, and the other lead to the respective Arduino input pin. Control Surface will enable the built-in pull-up resistors, so you don't have to add any resistors yourself.
2. Connect the buttons in a matrix configuration, follow the schematic posted above. In this case, you only connect the row pins and the column pins to the Arduino. Again, Control Surface will enable the built-in pull-up resistors, so don't add them externally.

Option 2 is pointless if you have 5 or fewer buttons. Matrices are useful for large numbers of pins.

Options 1 and 2 require different logic to read the state of each button, so your code needs to match your hardware configuration (e.g. use `NoteButton` for option 1 and `NoteButtonMatrix` for option 2).

**Reply by kentforth** [reply]
Thank you for helping me!
